### DISCUSSION TITLE
Security recommendations for Github Actions is insecure when combined with common practices for complex input types

### DISCUSSION THREAD

**Discussion by paulsavoie** [community/community#119934] [Actions] [answered]
### Select Topic Area

Product Feedback

### Body

**Problem statement:**
In the [GitHub documentation for handling user input securely](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#good-practices-for-mitigating-script-injection-attacks), it is noted as a good practice to pass variables using `with:` to a dedicated action, providing the sense that this handling of user input values is secure:

```
uses: fakeaction/checktitle@v3
with:
    title: ${{ github.event.pull_request.title }}
```

Input for GitHub Actions actions are currently limited to string types. If more complex types, such as dictionaries or arrays have to be provided, it is common practice to pass the complex input in some defined format and parse it in the action (see [here](https://stackoverflow.com/questions/75420197/how-to-use-array-input-for-a-custom-github-actions), [here](https://github.com/google-github-actions/get-secretmanager-secrets) and [here](https://github.com/aws-actions/configure-aws-credentials)). If combined with the recommendation mentioned above, this creates workflow steps as follows:

``` Example
uses: my-repo/my-action@v1
with:
   complex-param: >-
    {
     "arg1": "my-inline-secure-string",
     "arg2": "${{ vars.my-insecure-var }}"
    }
```

This sample, is however highly insecure, as a value such as `my-string", "arg1": "overwrite"` of  variable `vars.my-insecure-vars` would actually overwrite the value of `arg1`. If the input is not trusted, this can easily be overlooked and considered safe.

We recommend to at least include a case for complex objects in the documentation, highlighting the security implications, providing ways for input sanitization or even better, allowing complex objects (at least arrays) to be defined as inputs for a custom action.

Related discussions are [11692](https://github.com/orgs/community/discussions/11692), [25353](https://github.com/orgs/community/discussions/25353) and [32054](https://github.com/orgs/community/discussions/32054)

**Comment by ChristopherHX** [comment]
I would harden you example like this against json injection
```yaml
uses: my-repo/my-action@v1
with:
   complex-param: >-
    {
     "arg1": "my-inline-secure-string",
     "arg2": ${{ tojson(vars.my-insecure-var) }}
    }
```

I recommend to open an issue at github/docs, sometimes it can take longer without an pr to the same repo.

> allowing complex objects (at least arrays) to be defined as inputs for a custom action.

I would also like to be able to use complex object in composite actions and reusable workflows

**Reply by paulsavoie** [reply]
Hello @ChristopherHX , thank you for your message. I was not looking for a workaround for ourselves, we are aware of the issue and have mitigated it for our own action. 

I merely wanted to report that the security recommendations are misleading. Not everyone will think it through that such an attack can happen and that the variables have to be properly escaped. 

If it is necessary to harden the invocation of the action, then this should be explicitly stated in the security recommendations on the linked page. The current documentation creates a false sense of security and that, when following the practices, one's workflow definition is secure.

**Reply by ChristopherHX** [reply, marked as answer]
The chance for an respone from the gihub docs team still seems to be much higher for me on https://github.com/github/docs/issues and I still recommend to repost it there, some topics in the community forum run stale without response from github

**Reply by paulsavoie** [reply]
Thank you, I will post it there!
